cli/src/semgrep/rule_model.py

```diff
@@ -0,0 +1,1098 @@
+    search = "search"
+    taint = "taint"
+    _join = "join"  # NOTE: This is a workaround to avoid a name conflict with the str `join` field
+    extract = "extract"
+
+
+# NOTE: Duplication is better than the wrong abstraction here.
+# TODO: Figure out why JoinedRules supports this partial match of our Mode enum
+#       and provide a better solution.
+class JoinedRuleMode(str, Enum):
+    search = "search"
+    taint = "taint"
+
+
+# NOTE: Coupling with lang.json as rule_schema_v1.yaml is out-of-date
+class Languages(str, Enum):
+    aliengrep = "aliengrep"
+    apex = "apex"
+    bash = "bash"
+    c = "c"
+    cs = "c#"
+    cs2 = "C#"  # TODO: Normalize raw values to become case-insensitive
+    cpp_ = "c++"
+    cairo = "cairo"
+    clojure = "clojure"
+    cpp = "cpp"
+    csharp = "csharp"
+    dart = "dart"
+    docker = "docker"
+    dockerfile = "dockerfile"
+    elixir = "elixir"
+    ex = "ex"
+    generic = "generic"
+    go = "go"
+    golang = "golang"
+    hack = "hack"
+    hcl = "hcl"
+    html = "html"
+    java = "java"
+    javascript = "javascript"
+    js = "js"
+    json = "json"
+    jsonnet = "jsonnet"
+    julia = "julia"
+    kotlin = "kotlin"
+    kt = "kt"
+    lisp = "lisp"
+    lua = "lua"
+    move_on_aptos = "move-on-aptos"
+    none = "none"
+    ocaml = "ocaml"
+    php = "php"
+    promql = "promql"
+    proto = "proto"
+    proto3 = "proto3"
+    protobuf = "protobuf"
+    py = "py"
+    python = "python"
+    python2 = "python2"
+    python3 = "python3"
+    ql = "ql"
+    r = "r"
+    regex = "regex"
+    ruby = "ruby"
+    rust = "rust"
+    scala = "scala"
+    scheme = "scheme"
+    sh = "sh"
+    sol = "sol"
+    solidity = "solidity"
+    spacegrep = "spacegrep"
+    swift = "swift"
+    terraform = "terraform"
+    tf = "tf"
+    ts = "ts"
+    typescript = "typescript"
+    vue = "vue"
+    xml = "xml"
+    yaml = "yaml"
+
+
+def check_language_static(v: Union[str, List[str]]) -> Union[str, List[str]]:
+    if isinstance(v, list):
+        # Check for singleton list and return the string for a better error message
+        if len(v) == 1:
+            return v[0]
+        # Otherwise, check for invalid languages
+        _invalid_langs = [lang for lang in v if not hasattr(Languages, lang)]
+        # If there's only one invalid language, return it as a string for a better error message
+        invalid_langs = (
+            _invalid_langs[0]
+            if _invalid_langs and len(_invalid_langs) == 1
+            else _invalid_langs
+        )
+        if invalid_langs:
+            valid_langs = [f"'{lang.value}'" for lang in Languages]
+            if len(valid_langs) > 1:  # Add an "or" to the last valid language
+                valid_langs[-1] = f"or {valid_langs[-1]}"
+            expected_msg = f"Expected one of {', '.join(lang for lang in valid_langs)}"
+            raise ValueError(f"Invalid languages `{invalid_langs}`, {expected_msg}")
+    return v
+
+
+class Severity(str, Enum):
```

Comment:
Is this file auto generated?

Reply:
autogenerated and then hand-tuned.

Yes, we should use an RPC or perhaps the work @kopecs is in progress with in https://github.com/semgrep/semgrep-proprietary/pull/1898 will cover us here

Reply:
Using RPC for something we already do in OCaml seems a bit silly; it's not like this is something happening in either/or---it happens in both currently, and not for any great reason in the Python afaik (other than some error-message quality differences). I would sooner just delete the jsonschema.validate call rather than reimplement with pydantic (Do you have an example of a pydantic error message?).

I am not going to be working on that PR in the foreseeable future unless I can get time allocated for it away from JS-related work. Happy to help anyone get it merged if the outcomes are needed for another team.

Reply:
> and not for any great reason in the Python afaik (other than some error-message quality differences)
💯 this – I've wanted to [tear out](https://github.com/semgrep/semgrep/pull/9797/commits/5d6baec1c5c3a8a907788168e31c66cc1dce89b0) this `jsonschema.validate` call earlier in February when it added ~1.2s for logged out scans and ~4s for logged-in scans (time delta due to more rules to validate). Now, we're taking 8s for the `auto` rules and up to 40s for some scans I've seen in our jaeger traces.

I added some custom error messaging in the initial PR and scrubbed it here since there was too much test cruft to work through.